File: solutions/julia/circular-buffer/1/circular-buffer.jl
mutable struct CircularBuffer{T} <: AbstractVector{T}
    items::Vector{T}
    currentposition::Int
    capacity::Int

    CircularBuffer{T}(capacity::Integer) where {T} = new(zeros(T, capacity), 0, capacity)
end

function Base.push!(cb::CircularBuffer, item; overwrite::Bool=false)
    cb.currentposition == 
end

function Base.popfirst!(cb::CircularBuffer)

end

function Base.empty!(cb::CircularBuffer)

end

File: solutions/csharp/circular-buffer/1/CircularBuffer.cs
using System;
using System.Collections.Generic;

public class CircularBuffer<T>
{
    private Queue<T> _items;
    private int _count;
    private readonly int _capacity;

    public CircularBuffer(int capacity)
    {
        _items = new Queue<T>();
        _capacity = capacity;
    }

    public T Read()
    {
        if (_count == 0)
            throw new InvalidOperationException("Buffer is empty!");

        _count--;

        return _items.Dequeue();
    }

    public void Write(T value)
    {
        if (_count == _capacity)
            throw new InvalidOperationException("Buffer is full!");

        _items.Enqueue(value);
        _count++;
    }

    public void Overwrite(T value)
    {
        if (_count == _capacity)
            Read();

        Write(value);
    }

    public void Clear()
    {
        _items.Clear();
        _count = 0;
    }
}